program-runtime/src/loaded_programs.rs

```diff
@@ -721,33 +721,38 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
         }) {
             Ok(index) => {
                 let existing = slot_versions.get_mut(index).unwrap();
-                if std::mem::discriminant(&existing.program)
-                    != std::mem::discriminant(&entry.program)
-                {
-                    // Copy over the usage counter to the new entry
-                    entry.tx_usage_counter.fetch_add(
-                        existing.tx_usage_counter.load(Ordering::Relaxed),
-                        Ordering::Relaxed,
-                    );
-                    entry.ix_usage_counter.fetch_add(
-                        existing.ix_usage_counter.load(Ordering::Relaxed),
-                        Ordering::Relaxed,
-                    );
-                    *existing = entry.clone();
-                    self.stats.reloads.fetch_add(1, Ordering::Relaxed);
-                    false
-                } else {
-                    // Something is wrong, I can feel it ...
-                    self.stats.replacements.fetch_add(1, Ordering::Relaxed);
-                    true
+                match (&existing.program, &entry.program) {
+                    (LoadedProgramType::Builtin(_), LoadedProgramType::Builtin(_))
+                    | (LoadedProgramType::Unloaded(_), LoadedProgramType::LegacyV0(_))
+                    | (LoadedProgramType::Unloaded(_), LoadedProgramType::LegacyV1(_))
+                    | (LoadedProgramType::Unloaded(_), LoadedProgramType::Typed(_)) => {}
+                    #[cfg(test)]
+                    (LoadedProgramType::Unloaded(_), LoadedProgramType::TestLoaded(_)) => {}
```

Comment:
Can an `unloaded` entry move to `closed`?

Reply:
Program un/re/deployments change the deployment slot: https://github.com/solana-labs/solana/blob/31a73ab73177301af9d70ac0b57b94009dfdf678/programs/bpf_loader/src/lib.rs#L1134
Thus, it can not replace an existing entry.